src/Altinn.Apps/AppFrontend/react/altinn-app-frontend/src/components/base/CheckboxesContainerComponent.tsx

```diff
@@ -67,95 +68,46 @@ const useStyles = makeStyles({
+  const apiOptions: IOption[] = useAppSelector(state => state.optionState.options[props.optionsId]);
+  const options = apiOptions || props.options || emptyList;
   const checkBoxesIsRow: boolean = options.length <= 2;
 
-  React.useEffect(() => {
-    returnState();
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [options]);
+  const selected = React.useMemo(() => {
+    return props.formData.simpleBinding.split(',').map(v => options.find(o => o.value == v)).filter(o => !!o) ?? emptyList;
+  }, [props.formData, options])
 
+  // Implement preselected functionality
   React.useEffect(() => {
-    returnState();
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [props.formData?.simpleBinding]);
-
-  const returnState = () => {
+    const preselectedOption = options?.[props.preselectedOptionIndex]?.value;
     if (
-      !props.formData?.simpleBinding &&
-      props.preselectedOptionIndex >= 0 &&
-      options &&
-      props.preselectedOptionIndex < options.length
+      !selected &&
+      preselectedOption
     ) {
-      const preSelected: string[] = [];
-      preSelected[props.preselectedOptionIndex] =
-        options[props.preselectedOptionIndex].value;
-      props.handleDataChange(preSelected[props.preselectedOptionIndex]);
-      setSelected(preSelected);
-    } else {
-      setSelected(props.formData?.simpleBinding ? props.formData.simpleBinding.split(',') : []);
+      props.handleDataChange(preselectedOption);
     }
-  };
-
-  const onDataChanged = (event: React.ChangeEvent<HTMLInputElement>) => {
-    const newSelected: any = selected.slice();
-    const index = newSelected.indexOf(event.target.name);
+    // This effect should only run when new options are loaded
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [options]);
 
-    if (index >= 0) {
-      newSelected[index] = '';
-    } else {
-      newSelected.push(event.target.name);
-    }
-    const filtered = newSelected.filter((element: string) => !!element);
-    props.handleFocusUpdate(props.id);
-    props.handleDataChange(selectedHasValues(filtered) ? filtered.join() : '');
-  };
 
-  const handleOnBlur = () => {
-    props.handleDataChange(props.formData?.simpleBinding ?? '');
-  };
 
-  const selectedHasValues = (select: string[]): boolean => {
-    return select.some((element) => element !== '');
-  };
+  const onDataChanged = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const previouslySelected = selected.some(v => v.value === event.target.name);
 
-  const isOptionSelected = (option: string) => {
-    return selected.indexOf(option) > -1;
+    const newSelected = previouslySelected ?
+      selected.filter(o => o.value !== event.target.name) :
+      [...selected, options.find(o => o.value === event.target.name)];
+    props.handleDataChange(newSelected.map(o => o.value).join());
+    props.handleFocusUpdate(props.id);
   };
 
-  const inFocus = (index: number) => {
```

Comment:
I find it a bit difficult to follow all the logic here, and I'm not entirely sure what the details are around this, but from the looks of it, it seems like we are now missing some funcitonality for autofocus? The prop `shouldFocus` f.ex is not used anywhere.

Perhaps @lorang92 knows this area better and what is needed here?

Reply:
Yea, I didn't understand what this was doing either. The previous code mixes indexes in the `selected` array and the `options` array, so I'm pretty sure it is only correct by chance in some small special cases. 
![image](https://user-images.githubusercontent.com/131616/151002412-1c2e6914-2661-4af4-a676-865312c6f195.png)

Reply:
@ivarne / @haakemon i'm not familiar with this part of the code either. I can't really make sense of it. If regression tests run green my hunch here is that this could be removed.

Reply:
@jeevananthank just verify that this issue has not resurfaced when testing: https://github.com/Altinn/altinn-studio/issues/3782

Reply:
@ivarne ah, I see that you already mentioned that ❤️

Reply:
@lorang92 #3782 is not reproducible in the updated code